src/mlpack/core/optimizers/cne/cne_impl.hpp

```diff
@@ -0,0 +1,205 @@
+
+#include "cne.hpp"
+
+namespace mlpack {
+namespace optimization {
+
+CNE::CNE(const size_t populationSize,
+         const size_t maxGenerations,
+         const double mutationProb,
+         const double mutationSize,
+         const double selectPercent,
+         const double finalValue,
+         const double fitnessHist) :
+         populationSize(populationSize),
+         maxGenerations(maxGenerations),
+         mutationProb(mutationProb),
+         mutationSize(mutationSize),
+         selectPercent(selectPercent),
+         finalValue(finalValue),
+         fitnessHist(fitnessHist)
+         { /* Nothing to do here. */ }
+
+//! Optimize the function
+template<typename DecomposableFunctionType>
+double CNE::Optimize(
+    DecomposableFunctionType& function,
+    arma::mat& answer)
+{
+  // Get the number of functions to iterate
+  const size_t numFun = function.NumFunctions();
+
+  // Set the population size and fill random values [0,1]
+  population.set_size(populationSize, answer.n_rows);
+  population.randu();
+
+  // initialize helper variables
+  arma::vec parameters(answer.n_rows);
+  fitnessValues.set_size(populationSize);
+  double fitness = 0;
+  double lastBestFitness;
+
+  Log::Info << "CNE initialized successfully. Optimization started"
+  << std::endl;
+
+  // Iterate till max number of generations
+  for (size_t gen = 1; gen <= maxGenerations; gen++)
+  {
+    // calculate fitness values of all candidates
+    for (size_t i = 0; i < populationSize; i++)
+     {
+       // select a candidate
+       parameters = population.row(i).t();
+       // Insert the parameters in the function
+       answer = parameters;
+
+       // find the fitness
+       for (int j = 0; j < numFun; j++)
+       fitness += function.Evaluate(parameters, j);
+
+       // Save fitness values
+       fitnessValues[i] = fitness;
+       fitness = 0;
+     }
+
+      Log::Info << "Generation number: " << gen << " best fitness = "
+      << (double)fitnessValues.min() << std::endl;
+
+      // create the next generation of species
+      Reproduce();
+
+      // check for termination
+      if (finalValue >= (double)fitnessValues.min())
+        {
+          Log::Info << "Terminating. Given fitness criteria " << finalValue
+          << " > " << (double)fitnessValues.min() << std::endl;
+          break;
+        }
+
+      // check for termination
+      if (fitnessHist != DBL_MAX && gen != 1 &&
+        (lastBestFitness - (double)fitnessValues.min()) < fitnessHist)
+        {
+          Log::Info << "Terminating. Fitness History change "
+          << (lastBestFitness - (double)fitnessValues.min())
+          << " < " << fitnessHist << std::endl;
+          break;
+        }
+
+      // Store the best fitness of this generation before update
+      lastBestFitness = (double)fitnessValues.min();
+  }
+
+  // Set the best candidate into the network
+  answer = population.submat(index[0], 0, index[0], populationSize-1);
```

Comment:
Pedantic style issue; missing space between ``populationSize``, ``-`` and ``1``.